src/sync.h

```diff
@@ -397,4 +398,77 @@ class CSemaphoreGrant
     }
 };
 
+/**
+ * A generic container that protects an owned object with a mutex.
+ * This is like a mutex with an extension - owning an object of arbitrary type
+ * and providing a dereference operation which gives access to the owned object
+ * and transparently locks the mutex for the duration of the access. For example:
+ *
+ * @code{.cpp}
+ * Synced<std::vector<int>> v{4, 5, 6};
+ *
+ * std::thread t1{[&v]() { SYNCED_LOCK(v, p); p->push_back(7); }};
+ * std::thread t2{[&v]() { SYNCED_LOCK(v, p); p->push_back(8); }};
+ *
+ * SYNCED_LOCK(v, p);
+ * const size_t size = p->size();
+ * for (auto& i : *p) {
+ *     i *= 10;
+ * }
+ * // The container will not be changed by another thread between the two size() calls.
+ * assert(size == p->size());
+ *
+ * std::cout << (*p)[0] << std::endl;
+ *
+ * @endcode
+ */
+template <typename T>
+class Synced
+{
+public:
+    /**
+     * Construct the container and its internal object of type T,
+     * passing `args...` to T's constructor.
+     */
+    template <typename... Args>
+    Synced(Args... args) : m_obj{args...}
+    {
+    }
+
+    /**
+     * An auxiliary class that holds the parent locked for the duration of its
+     * lifetime and provides access to the internal object.
+     */
+    class SCOPED_LOCKABLE Proxy : public ::UniqueLock<Mutex>
+    {
+    public:
+        Proxy(Synced& parent LIFETIMEBOUND, const char* name, const char* file, int line)
+            EXCLUSIVE_LOCK_FUNCTION(parent.m_mutex)
+            : ::UniqueLock<Mutex>{parent.m_mutex, name, file, line}, m_parent{parent}
+        {
+        }
+
+        ~Proxy() UNLOCK_FUNCTION() {}
+
+        T& operator*() LIFETIMEBOUND { return m_parent.m_obj; }
+
+        T* operator->() LIFETIMEBOUND { return &m_parent.m_obj; }
+
+    private:
+        Synced& m_parent;
+    };
+
+protected:
+    mutable Mutex m_mutex;
+    T m_obj GUARDED_BY(m_mutex);
+};
+
+#define SYNCED_LOCK(synced, name) decltype(synced)::Proxy name{synced, #synced, __FILE__, __LINE__}
```

Comment:
Sometimes the hardcoded name could become too long/verbose for a too short scope: e.g. `g_reachable_networks_proxy` or `g_reachable_networks_locked` when it is just used on the line below and nowhere else.

I added now the possibility to call `SYNCED_LOCK()` with 1 or 2 arguments and opted for a `_locked` suffix instead of `_proxy` for the 1-arg case. So the callers have complete flexibility on how to call it.

Reply:
That does not compile on windows: https://cirrus-ci.com/task/4559637407072256?logs=build#L814 :-/

If you are ok with this, I will fall back to the 2-arg macro. Then the caller can pick the name depending on their preferences - `p`, `foo_proxy`, `foo_locked` or whatever.

Reply:
If you want naming it to be optional, probably better to have different macros, same as `LOCK(m)` and `WAIT_LOCK(m,g)`. I'm not really sure what I prefer here.

Reply:
I went back to what it was before - `SYNCED_LOCK(object, name)`. The 1-arg macro was used in just one place and it did not compile on windows, not worth it. By selecting the name the caller can choose whether to use a short one for small scopes of a few lines, or a longer and more descriptive one for bigger scopes.